service/src/main/java/bio/terra/workspace/common/utils/GcpUtils.java

```diff
@@ -65,22 +67,47 @@ public static Optional<Project> retrieveProject(
   }
 
   /**
-   * Poll until the Google Service API operation has completed. Throws any error or timeouts as a
+   * Poll until the Google Service API operation has completed. If an error is retryable, throws a
    * {@link RetryException}.
    */
-  public static void pollUntilSuccess(
+  public static void pollAndRetry(
       OperationCow<?> operation, Duration pollingInterval, Duration timeout)
       throws RetryException, IOException, InterruptedException {
     operation = OperationUtils.pollUntilComplete(operation, pollingInterval, timeout);
     if (operation.getOperationAdapter().getError() != null) {
-      throw new RetryException(
-          String.format(
-              "Error polling operation. name [%s] message [%s]",
-              operation.getOperationAdapter().getName(),
-              operation.getOperationAdapter().getError().getMessage()));
+      int code = operation.getOperationAdapter().getError().getCode();
+      if (is4xxClientError(code)) {
+        // do not waste time retrying on client error.
+        throw new InternalServerErrorException(
```

Comment:
Is that the right exception? That is a 5xx. If it is the WSM caller that passed in a bad VM, is that really an internal error?

Also, will this fix [PF-2042](https://broadworkbench.atlassian.net/browse/PF-2042)?

Reply:
ah right. hmm can i just throw a bad request then?

 not aware of that ticket. i can test it.

Reply:
If WSM is passing things through, then I would throw a BadRequest-class message. It is our caller that made the mistake and we need to tell them.

If WSM is checking the inputs and making sure we never pass in something bad, then InternalServiceError seems right.

I'm thinking BadRequest is better for this situation.

Reply:
it does fix that ticket. i passed in imageName `r-cpu.4-1:m95` and got a statusCode 3 ILLEGAL_ARGUMENT.

```
workspaceUuid: f6daaff5-1ba9-4bab-8b68-c52c8c5cc463
action: create_controlled_user_private
2022-12-15 15:56:09.654 eyWNwyN5  INFO 53596 --- [nio-8080-exec-7] bio.terra.workspace.db.WorkspaceDao      : Retrieved workspace record Workspace[workspaceId=f6daaff5-1ba9-4bab-8b68-c52c8c5cc463, userFacingId=testworkspace3, displayName=null, description=null, spendProfileId=bio.terra.workspace.service.spendprofile.SpendProfileId@e2ed6c6c, properties={}, workspaceStage=MC_WORKSPACE, createdByEmail=yuhuyoyo@google.com, createdDate=2022-12-14T19:37:10.104497Z]
2022-12-15 15:56:09.805 eyWNwyN5  INFO 53596 --- [nio-8080-exec-7] b.t.workspace.service.iam.SamService     : User yuhuyoyo@google.com is authorized to create_controlled_user_private workspace f6daaff5-1ba9-4bab-8b68-c52c8c5cc463
2022-12-15 15:56:09.957 eyWNwyN5  INFO 53596 --- [nio-8080-exec-7] bio.terra.workspace.db.WorkspaceDao      : Retrieved workspace record Workspace[workspaceId=f6daaff5-1ba9-4bab-8b68-c52c8c5cc463, userFacingId=testworkspace3, displayName=null, description=null, spendProfileId=bio.terra.workspace.service.spendprofile.SpendProfileId@e2ed6c6c, properties={}, workspaceStage=MC_WORKSPACE, createdByEmail=yuhuyoyo@google.com, createdDate=2022-12-14T19:37:10.104497Z]
2022-12-15 15:56:09.961 eyWNwyN5  INFO 53596 --- [nio-8080-exec-7] b.terra.workspace.common.utils.MdcHook   : Flight ID e2183bd0-7cba-11ed-a1eb-0242ac120002 changed status to RUNNING.
2022-12-15 15:56:09.961 eyWNwyN5  INFO 53596 --- [nio-8080-exec-7] bio.terra.stairway.impl.StairwayImpl     : Launching flight class: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight stairway: stairwaydab9e5c7-98ee-4b20-bfcb-a0835fdf41d8 flightid: e2183bd0-7cba-11ed-a1eb-0242ac120002
2022-12-15 15:56:09.962   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startFlight, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002
2022-12-15 15:56:09.964   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.StoreMetadataStep, stepIndex: 0, direction: DO
2022-12-15 15:56:09.968 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] bio.terra.workspace.db.ResourceDao       : Inserted record for resource 2478fc72-7cba-11ed-a1eb-0242ac120002 for workspace f6daaff5-1ba9-4bab-8b68-c52c8c5cc463
2022-12-15 15:56:09.968 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.StoreMetadataStep, stepIndex: 0, direction: DO
2022-12-15 15:56:09.970   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateSamResourceStep, stepIndex: 1, direction: DO
2022-12-15 15:56:10.061 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.t.workspace.service.iam.SamService     : Sam API error while creating a controlled resource, code is 409
2022-12-15 15:56:10.061 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.t.workspace.service.iam.SamService     : Sam error was CONFLICT on creation request. This means the resource already exists but is not an error so no exception thrown.
2022-12-15 15:56:10.061 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateSamResourceStep, stepIndex: 1, direction: DO
2022-12-15 15:56:10.063   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.GetCloudContextStep, stepIndex: 2, direction: DO
2022-12-15 15:56:10.065 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.GetCloudContextStep, stepIndex: 2, direction: DO
2022-12-15 15:56:10.068   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.RetrieveNetworkNameStep, stepIndex: 3, direction: DO
2022-12-15 15:56:10.843 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.RetrieveNetworkNameStep, stepIndex: 3, direction: DO
2022-12-15 15:56:10.846   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.GrantPetUsagePermissionStep, stepIndex: 4, direction: DO
2022-12-15 15:56:10.974 eyWNwyN5  INFO 53596 --- [nio-8080-exec-7] b.t.common.logging.RequestLoggingFilter  : POST /api/workspaces/v1/f6daaff5-1ba9-4bab-8b68-c52c8c5cc463/resources/controlled/gcp/ai-notebook-instances 202
2022-12-15 15:56:11.934 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.GrantPetUsagePermissionStep, stepIndex: 4, direction: DO
2022-12-15 15:56:11.938   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.CreateAiNotebookInstanceStep, stepIndex: 5, direction: DO
2022-12-15 15:56:32.726 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] bio.terra.stairway.impl.FlightRunner     : Caught exception: (bio.terra.common.exception.BadRequestException@5c33bb10[causes=[],statusCode=400 BAD_REQUEST]) flight id: e2183bd0-7cba-11ed-a1eb-0242ac120002 step: 5 direction: DO

bio.terra.common.exception.BadRequestException: gcp calls failed with client error code 3. Do not retry
        at bio.terra.workspace.common.utils.GcpUtils.pollAndRetry(GcpUtils.java:82)
        at bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.CreateAiNotebookInstanceStep.doStep(CreateAiNotebookInstanceStep.java:144)
        at bio.terra.stairway.impl.FlightRunner.stepWithRetry(FlightRunner.java:248)
        at bio.terra.stairway.impl.FlightRunner.runSteps(FlightRunner.java:179)
        at bio.terra.stairway.impl.FlightRunner.fly(FlightRunner.java:110)
        at bio.terra.stairway.impl.FlightRunner.run(FlightRunner.java:68)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)

2022-12-15 15:56:32.727 eyWNwyN5 ERROR 53596 --- [pool-7-thread-3] bio.terra.stairway.StepResult            : Step Result exception

bio.terra.common.exception.BadRequestException: gcp calls failed with client error code 3. Do not retry
        at bio.terra.workspace.common.utils.GcpUtils.pollAndRetry(GcpUtils.java:82)
        at bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.CreateAiNotebookInstanceStep.doStep(CreateAiNotebookInstanceStep.java:144)
        at bio.terra.stairway.impl.FlightRunner.stepWithRetry(FlightRunner.java:248)
        at bio.terra.stairway.impl.FlightRunner.runSteps(FlightRunner.java:179)
        at bio.terra.stairway.impl.FlightRunner.fly(FlightRunner.java:110)
        at bio.terra.stairway.impl.FlightRunner.run(FlightRunner.java:68)
        at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:539)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)

2022-12-15 15:56:32.727 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.CreateAiNotebookInstanceStep, stepIndex: 5, direction: DO
2022-12-15 15:56:32.734   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.CreateAiNotebookInstanceStep, stepIndex: 5, direction: SWITCH
2022-12-15 15:56:32.962 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.CreateAiNotebookInstanceStep, stepIndex: 5, direction: SWITCH
2022-12-15 15:56:32.968   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.GrantPetUsagePermissionStep, stepIndex: 4, direction: UNDO
2022-12-15 15:56:33.799 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.GrantPetUsagePermissionStep, stepIndex: 4, direction: UNDO
2022-12-15 15:56:33.806   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.RetrieveNetworkNameStep, stepIndex: 3, direction: UNDO
2022-12-15 15:56:33.807 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.cloud.gcp.ainotebook.RetrieveNetworkNameStep, stepIndex: 3, direction: UNDO
2022-12-15 15:56:33.810   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.GetCloudContextStep, stepIndex: 2, direction: UNDO
2022-12-15 15:56:33.811 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.GetCloudContextStep, stepIndex: 2, direction: UNDO
2022-12-15 15:56:33.813   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateSamResourceStep, stepIndex: 1, direction: UNDO
2022-12-15 15:56:33.871 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.t.workspace.service.iam.SamService     : Sam API error while deleting a controlled resource, code is 404
2022-12-15 15:56:33.871 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.t.workspace.service.iam.SamService     : Sam error was NOT_FOUND on a deletion call. This just means the deletion was tried twice so no error thrown.
2022-12-15 15:56:33.871 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateSamResourceStep, stepIndex: 1, direction: UNDO
2022-12-15 15:56:33.875   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: startStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.StoreMetadataStep, stepIndex: 0, direction: UNDO
2022-12-15 15:56:33.882 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] bio.terra.workspace.db.ResourceDao       : Deleted record for resource 2478fc72-7cba-11ed-a1eb-0242ac120002 in workspace f6daaff5-1ba9-4bab-8b68-c52c8c5cc463
2022-12-15 15:56:33.883 eyWNwyN5  INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endStep, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002, stepClass: bio.terra.workspace.service.resource.controlled.flight.create.StoreMetadataStep, stepIndex: 0, direction: UNDO
2022-12-15 15:56:33.886   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Flight ID e2183bd0-7cba-11ed-a1eb-0242ac120002 changed status to ERROR.
2022-12-15 15:56:33.887   INFO 53596 --- [pool-7-thread-3] b.terra.workspace.common.utils.MdcHook   : Operation: endFlight, flightClass: bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight, flightId: e2183bd0-7cba-11ed-a1eb-0242ac120002
2022-12-15 15:56:33.887   INFO 53596 --- [pool-7-thread-3] bio.terra.stairway.StairwayHook          : endFlight bio.terra.workspace.service.resource.controlled.flight.create.CreateControlledResourceFlight: ERROR
```

Reply:
Yay! Thanks for checking.